### PR TITLE
feat(generator): per-method proxy override on generated services

### LINKED ISSUE #117: generator: add per-method `proxy: String? = null` to generated service queries / procedures
## Context

`io.github.kikin81.atproto.runtime.XrpcClient.{query, procedure}` already accept `proxy: String? = null` and forward the value as the `atproto-proxy` header. The generated service classes wrap those calls but only expose `proxy` when an NSID prefix has a hardcoded mapping in [`generator/src/main/kotlin/io/github/kikin81/atproto/generator/emit/ProxyMapping.kt`](https://github.com/kikin81/atproto-kotlin/blob/main/generator/src/main/kotlin/io/github/kikin81/atproto/generator/emit/ProxyMapping.kt). Today that table is:

```kotlin
private val rules: List<Pair<String, String>> = listOf(
    "chat.bsky." to "did:web:api.bsky.chat#bsky_chat",
)
```

For `chat.bsky.*` this works perfectly — every NSID routes to the same fixed appview, so [`ServiceGenerator`](https://github.com/kikin81/atproto-kotlin/blob/main/generator/src/main/kotlin/io/github/kikin81/atproto/generator/emit/ServiceGenerator.kt) emits a `proxy: String? = \"did:web:api.bsky.chat#bsky_chat\"` constructor parameter on `ConvoService` and forwards `proxy = proxy` from every method body. ✅

The gap: **NSIDs whose proxy target is consumer-chosen, not lexicon-fixed.** The canonical example is `app.bsky.notification.{registerPush, unregisterPush}`. The wire contract requires the `atproto-proxy: <gateway-did>#bsky_notif` header so the user's PDS forwards the call to the push gateway — but **the gateway DID is a consumer choice**, not a Bluesky-managed appview address. Different consumers point at different gateways:

- Bluesky's official client → `did:web:api.bsky.app#bsky_notif`
- Self-hosted gateways → e.g. `did:web:push.nubecita.app#bsky_notif`, or any other operator's DID

So `ProxyMapping` can't (and shouldn't) hardcode it. But there's currently no escape hatch: `NotificationService.registerPush(request)` has no `proxy` parameter, and the generated body calls `client.procedure(...)` without forwarding one.

## Concrete consumer pain

[`nubecita`'s `:core:push` module (PR #301)](https://github.com/kikin81/nubecita/pull/301) had to bypass the generated wrapper entirely:

```kotlin
// core/push/src/main/kotlin/net/kikin/nubecita/core/push/DefaultPushRegistrationRepository.kt
override suspend fun register(did: String, fcmToken: String): Result<Unit> = runCatching {
    val client = xrpcClientProvider.authenticated()
    client.procedure(
        nsid = \"app.bsky.notification.registerPush\",
        params = NoXrpcParams,
        paramsSerializer = NoXrpcParams.serializer(),
        input = RegisterPushRequest(
            serviceDid = Did(\"did:web:push.nubecita.app\"),
            token = fcmToken,
            platform = \"android\",
            appId = appId,
        ),
        inputSerializer = RegisterPushRequest.serializer(),
        responseSerializer = UnitResponseSerializer,
        proxy = \"did:web:push.nubecita.app#bsky_notif\",
    )
}
```

The typed DTOs (`RegisterPushRequest`, `UnregisterPushRequest`) are perfectly usable — we just can't get to them through the generated service wrapper. The bypass duplicates the wrapper's NSID string, serializer wiring, and response handling at every call site, and consumers lose the ergonomics of `notificationService.registerPush(request, proxy = …)`.

## Proposed change

Have `ServiceGenerator` emit a **per-method** `proxy: String? = null` parameter on every generated query/procedure method, in addition to (not instead of) the existing constructor-level `proxy` property. The method-level parameter wins when non-null; falls back to the constructor-level proxy when null. Concretely, the generated method body becomes:

```kotlin
public suspend fun registerPush(
    request: RegisterPushRequest,
    proxy: String? = null,
): Unit = client.procedure(
    nsid = \"app.bsky.notification.registerPush\",
    params = NoXrpcParams,
    paramsSerializer = NoXrpcParams.serializer(),
    input = request,
    inputSerializer = RegisterPushRequest.serializer(),
    responseSerializer = UnitResponseSerializer,
    proxy = proxy ?: this.proxy, // \"this.proxy\" only present when ProxyMapping has a hit
)
```

For services with a `ProxyMapping` entry (`ConvoService` today):

- `convoService.listConvos(params)` — uses constructor default `did:web:api.bsky.chat#bsky_chat`, identical to today.
- `convoService.listConvos(params, proxy = \"did:web:custom.chat#bsky_chat\")` — caller override for self-hosted chat appviews. (Not a use case today, but a free win.)

For services without a `ProxyMapping` entry (`NotificationService`, all `app.bsky.*` repos, etc.):

- `notificationService.registerPush(request)` — `proxy = null`, identical to today (no header).
- `notificationService.registerPush(request, proxy = \"did:web:push.nubecita.app#bsky_notif\")` — **new**, unblocks the consumer-chosen-proxy case.

The constructor-level `proxy` property generation logic doesn't change; we just add a new method-level parameter and adjust the forwarding expression.

### Files to touch

- [`generator/src/main/kotlin/io/github/kikin81/atproto/generator/emit/ServiceGenerator.kt`](https://github.com/kikin81/atproto-kotlin/blob/main/generator/src/main/kotlin/io/github/kikin81/atproto/generator/emit/ServiceGenerator.kt) — `buildQueryMethod` / `buildProcedureMethod` (~lines 364/419 emit the current `proxy = proxy` line; needs to become `proxy = proxy ?: this.proxy` when `emitProxy` is true, or just `proxy = proxy` when there's no service-level proxy).
- Golden test snapshots under `generator/src/test/.../golden/` will need regeneration; the `GoldenFileTest` flow already exists.

### Tests

- Extend [`ServiceGeneratorResolveProxyTest.kt`](https://github.com/kikin81/atproto-kotlin/blob/main/generator/src/test/kotlin/io/github/kikin81/atproto/generator/emit/ServiceGeneratorResolveProxyTest.kt) to assert the method-level `proxy: String? = null` param is emitted on every query/procedure, regardless of whether the service has a `ProxyMapping` hit.
- Add a test that the forwarding expression collapses correctly: when method-level `proxy` is non-null it wins; when null, it falls back to the constructor-level proxy (for `chat.bsky.*`) or is `null` (for `app.bsky.*`).
- Regenerate the smoke fixture (`generator/src/test/.../smoke/`) and confirm `NotificationService` ends up with `registerPush(request, proxy: String? = null)`.

## Acceptance

```kotlin
// nubecita's :core:push, post-upgrade — replaces the entire XrpcClient.procedure bypass
override suspend fun register(did: String, fcmToken: String): Result<Unit> = runCatching {
    val service = NotificationService(xrpcClientProvider.authenticated())
    service.registerPush(
        request = RegisterPushRequest(
            serviceDid = Did(\"did:web:push.nubecita.app\"),
            token = fcmToken,
            platform = \"android\",
            appId = appId,
        ),
        proxy = \"did:web:push.nubecita.app#bsky_notif\",
    )
}
```

…and the analogous `unregister` call.

## Out of scope (file separately if needed)

- **`NoBodyResponseSerializer`** — a friendlier marker than `UnitResponseSerializer` for procedures with no `output.schema`. `UnitResponseSerializer` requires the response body to be `{}` (fails on empty `\"\"`). Real PDSes / appviews send `{}` in practice but it tripped nubecita's `MockEngine`-based tests until we matched that. Cosmetic, not blocking.
- **Lexicon-declared proxy hints.** If a future AT Protocol revision adds routing metadata to the lexicon JSON, `ProxyMapping` could read it instead of the hardcoded table. Not blocking anything today.

## Refs

- Consumer PR: https://github.com/kikin81/nubecita/pull/301
- Prior chat-proxy work (resolved via `ProxyMapping`): https://github.com/kikin81/atproto-kotlin/issues/86
- Push gateway: [DracoBlue/atproto-push-gateway](https://github.com/DracoBlue/atproto-push-gateway) v1.2.0 deployment at `https://push.nubecita.app`

### PR DESCRIPTION
**Pull request by kikin81**
## Summary

- Every generated `*Service.<method>(...)` in `:models` now ends with an optional `proxy: String? = null` parameter. When non-null it overrides the service's constructor-level `proxy` default (if any) for that single call; when null the call is wire-identical to v8.
- Driving use case: `app.bsky.notification.{registerPush, unregisterPush}` where the gateway DID is consumer-chosen (e.g. `did:web:push.nubecita.app#bsky_notif`). Until now consumers had to bypass the typed wrapper — this restores the typed call site. See [nubecita#301](https://github.com/kikin81/nubecita/pull/301) for the consumer pain.
- Forwarding semantics: `proxy = proxy ?: this.proxy` when the service has a constructor `proxy` (today: `ConvoService`), `proxy = proxy` otherwise. `ProxyMapping`, `resolveProxyForPackage`, and `:runtime` are unchanged.

**BREAKING CHANGE** — Every generated method's JVM signature gained a `Ljava/lang/String;` parameter slot (and a `$default` bridge). Kotlin callers see no source-level break — defaults preserve v8 wire behavior — but binary consumers must recompile. semantic-release should cut **v9.0.0** on merge. Full migration notes in [`docs/breaking-changes/v9.md`](docs/breaking-changes/v9.md).

OpenSpec change: [`openspec/changes/add-per-method-proxy-override/`](openspec/changes/add-per-method-proxy-override/) (proposal, design, lexicon-codegen delta spec, tasks).

Closes #117.
Tracks bd kikinlex-b1h.

## Reviewer notes

The goldens diff is uniform — every emitted method gets the new parameter and forwarding line. Three services in the test corpus (`ConvoService`, `ExampleService`, `FeedService`) cover proxied / unproxied / raw-bytes shapes. `models/api/models.api` is the full surface impact across the real lexicon corpus.

## Test plan

- [x] `./gradlew :generator:test` — 93 tests pass (includes new `ServiceGeneratorProxyEmissionTest`, untouched `ServiceGeneratorResolveProxyTest`, and the regenerated `GoldenFileTest`)
- [x] `./gradlew :models:assemble` + `:models:apiDump` + `:models:apiCheck` — regenerated `NotificationService.registerPush/unregisterPush` and `ConvoService.listConvos` carry the new parameter; refreshed `models/api/models.api` passes apiCheck
- [x] `./gradlew build` — 383 tasks, all green; spotless clean
- [x] Smoke fixture spot-check — `NotificationService.registerPush(request, proxy: String? = null)` with `proxy = proxy` forwarding
- [x] `runtime/api/runtime.api` and `oauth/api/oauth.api` untouched

## Follow-ups (out of scope per #117)

- bd `kikinlex-l7u` — `NoBodyResponseSerializer` marker for procedures with no `output.schema`
- bd `kikinlex-90j` — lexicon-declared `atproto-proxy` hints (replace hardcoded `ProxyMapping`)

### CODE REVIEW

File: openspec/changes/add-per-method-proxy-override/design.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+Method parameter `proxy` shadows the class property of the same name, so `this.proxy` is the only way to reference the constructor-level value. Kotlin's Elvis operator collapses to the method override when non-null, the constructor default otherwise — exactly the semantics we want, in one line.
+
+The alternative — emit two overloads (one without `proxy`, one with) — doubles the surface area, fights against KotlinPoet's parameter-with-default model, and gains nothing.
+
+For services without a constructor `proxy`, the expression collapses to just `proxy = proxy`. No `this.proxy` reference; nothing to fall back to.
+
+### Decision 3: Cut v9.0.0; no compatibility shim
+
+Adding any parameter to a public API tracked by `kotlinx-binary-compatibility-validator` is binary-breaking even with a default value — captured in bd memory `adding-any-parameter-to-a-public-api-xrpcclient`, learned from prior breaking releases. The validator runs as `:models:apiCheck` in CI; the only way to satisfy it is to refresh `models/api/models.api` and accept the major bump.
```
[the remark this answers is not in the document]
Comment (kikin81) [reply openspec/changes/add-per-method-proxy-override/design.md]: Good catch. Rewrote this in 9820b915 — the design now points at the actual repo gate (`:models:apiCheck` in CI) and the prior `vN.md` breaking-changes precedent, instead of an internal session-memory key that nobody else can see. Self-contained for future readers.

File: openspec/changes/add-per-method-proxy-override/design.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+Alternatives rejected:
+- **Hidden internal flag** to skip parameter emission on a per-package basis: defeats the entire point of the change (uniformity).
+- **New service variant** (`NotificationServiceWithProxy`): API duplication, naming bikeshed, doesn't generalize.
+- **Defer the bump by emitting two separate functions** (one wrapping the other): same binary breakage in the underlying generated method, just buried.
+
+### Decision 4: Test surface — full plan→emit, not isolated unit
+
+The existing `ServiceGeneratorResolveProxyTest` tests the pure function `resolveProxyForPackage` in isolation — it's not appropriate for asserting emitted Kotlin syntax. The new test runs a small synthetic lexicon corpus (one `chat.bsky.*` NSID, one `app.bsky.*` NSID, both with at least one query and one procedure) through `LexiconParser → RefResolver → ContextTagger → NamingMatrix → VerificationPass → EmissionPlan → ServiceGenerator`, then asserts against the emitted `TypeSpec`:
```
[the remark this answers is not in the document]
Comment (kikin81) [reply openspec/changes/add-per-method-proxy-override/design.md]: Valid — doc and code disagreed. Updated in 9820b915 to match what `ServiceGeneratorProxyEmissionTest` actually does (`LexiconParser → SymbolTable.build → RefResolver.validate → ContextTagger → EmissionPlan.build → ServiceGenerator.emitAll`) and called out why `VerificationPass` is intentionally omitted: it runs *after* `ServiceGenerator.emitAll` to catch naming collisions across the full corpus, but the synthetic 7-lexicon corpus is too small to trip one, and the test's invariants are about per-method parameter and body shape rather than naming.
